shared/js/background/storage/https.es6.js

```diff
@@ -0,0 +1,71 @@
+const load = require('./../load.es6.js')
+const Dexie = require('dexie')
+const constants = require('../../../data/constants')
+
+class HTTPSStorage {
+    constructor () {
+        this.dbc = new Dexie('https')
+        this.dbc.version(1).stores({
+            httpsStorage: 'name,type,data'
+        })
+    }
+
+    // gets list data, returns array of promises. use promise.all().then()
+    getLists () {
+        let promiseList = []
+
+        constants.httpsLists.forEach((list) => {
+            promiseList.push(new Promise((resolve, reject) => {
+                this.getDataXHR(list.url).then(data => {
+                    if (data) {
+                        // if we have new data store it in local DB for later
+                        this.storeInLocalDB(list.name, list.type, data)
+                        list.data = data
+                        resolve(list)
+                    } else {
+                        // No new data, look up old data from DB
+                        this.getDataFromLocalDB(list.name).then(storedData => {
+                            if (storedData.data) {
+                                list.data = storedData.data
+                            }
+                            resolve(list)
+                        })
+                    }
+                })
```

Comment:
`this.getDataXHR` will never resolve if:
- request is unsuccessful or 
- request takes forever.

This means `getLists` never resolves, which means that `HTTPS.setLists` is never called leaving user (unconsciously) unprotected.

https://github.com/duckduckgo/duckduckgo-privacy-extension/blob/efc3ed53117fbb2d8886038be6978ff5b19cce10/shared/js/background/https.es6.js#L55-L56

This is potential vector of the attack - all attacker has to do is to block (or postpone indefinitely) request to `duckduckgo.com/https-bloom.json` or `duckduckgo.com/whitelist.json`.

We should do couple of things:
- load local list as soon as possible, don't wait for update (update only once we get new, valid data)
- have a timeout on https://github.com/duckduckgo/duckduckgo-privacy-extension/blob/efc3ed53117fbb2d8886038be6978ff5b19cce10/shared/js/background/load.es6.js#L31
- don't resolve only on success - https://github.com/duckduckgo/duckduckgo-privacy-extension/blob/efc3ed53117fbb2d8886038be6978ff5b19cce10/shared/js/background/load.es6.js#L50-L51
- add missing `.catch`es
- potentially inform user if we are unable to load local filter/list or download updated list for a long time
- I also think that HTTPS object should not be responsible for updating lists, there should be a separate entity for that. HTTPS should only load local files while other entity regularly updates lists and saves them locally.

Reply:
I generally agree with all those points. Here's a few notes:

> don't resolve only on success

FWIW I've added some error handling for the `load` module here #238 so we'd be able to tackle different error states after that PR is merged.

> have a timeout

Shouldn't the XHR fail eventually on timeout?

> This is potential vector of the attack - all attacker has to do is to block (or postpone indefinitely) request to `duckduckgo.com/https-bloom.json` or `duckduckgo.com/whitelist.json`.

Can you elaborate on the attack vector piece? We've got HSTS on duckduckgo.com and it's really unlikely that a user will install the extension without having visited DDG immediately before. Once the user downloads the HTTPS files for the first time, they're covered.

> I also think that HTTPS object should not be responsible for updating lists, there should be a separate entity for that. HTTPS should only load local files while other entity regularly updates lists and saves them locally.

This *is* the separate entity, we're separating storage from business logic in this PR :)

Reply:
> FWIW I've added some error handling for the load module here #238 so we'd be able to tackle different error states after that PR is merged.

Awesome 👏 That almost fixes the `request is unsuccessful` issue I mentioned, we just need to add missing `.catch`es to this file.

> Shouldn't the XHR fail eventually on timeout?

Unfortunately, not. By default there is no timeout. https://stackoverflow.com/questions/9502410/ajax-xmlhttprequest-timeout-length-by-browser#9503384

> Once the user downloads the HTTPS files for the first time, they're covered.

Blocking `duckduckgo.com/https-bloom.json` will cause HTTPS object to never enter the `isReady` state (as far as I can see), which completely disables its functionality. Even though the (older versions) of those lists are available locally.

Reply:
> Unfortunately, not. By default there is no timeout.

Huh, TIL. Yeah, let's add a timeout.

> Blocking `duckduckgo.com/https-bloom.json` will cause HTTPS object to never enter the `isReady` state (as far as I can see), which completely disables its functionality. Even though the (older versions) of those lists are available locally.

Got it. We should handle this here, yeah.

@jdorweiler this actually works pretty well with the thing I was unsure about in the other PR, where [the load promise gets rejected if the server returns 304](https://github.com/duckduckgo/duckduckgo-privacy-extension/pull/238/files#r199819195). In case of any kind of rejected promise from the `load` module, we can fall back to the DB.